Usage price override

sruthipendyala opened the pull request:
Initial Implementation of usage price override. Started off with implementing only Tiered Block Price Override, for Usage CONSUMABLE IN ARREAR.

Usage price override #614

sruthipendyala commented on catalog/src/main/java/org/killbill/billing/catalog/override/DefaultPriceOverride.java:
```diff
@@ -97,12 +111,98 @@ public boolean apply(final PlanPhasePriceOverride input) {
+                  List<TierPriceOverride>  tierPriceOverrides = getResolvedTierOverrides(curUsage.getTiers(), curOverride.getTierPriceOverrides());
+                  resolvedUsageOverrides.add(new DefaultUsagePriceOverride(curUsage.getName(), curUsage.getUsageType(),tierPriceOverrides));
+              } else
+                 resolvedUsageOverrides.add(null);
```
Adding null to the list as a placeholder to usage which does not have any override
Please suggest me if there is any better way to implement it.

sbrossie replied:
Could you instead return the original non overridden entry?

sbrossie commented:
@sruthipendyala You will need to first resolve the conflicts before we can merge.

sruthipendyala commented:
@sbrossie  sorry about that! fixed the merge conflicts

pierre commented on catalog/src/main/java/org/killbill/billing/catalog/caching/EhCacheOverriddenPlanCache.java:
```diff
@@ -91,28 +102,113 @@ private DefaultPlan loadOverriddenPlan(final String planName, final StaticCatalo
+              List<UsagePriceOverride> usagePriceOverrides =  getUsagePriceOverrides(curPhase, overriddenPhase, context);
+              result[i] = new DefaultPlanPhasePriceOverride(curPhase.getName(), Currency.valueOf(overriddenPhase.getCurrency()), overriddenPhase.getFixedPrice(), overriddenPhase.getRecurringPrice(), usagePriceOverrides);
+            }
+            else
```
Nit: we prefer to always explicitly add the curly braces (same remark below).

pierre commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/json/TierpRICEOverrideJson.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public class TierPriceOverrideJson {
```
⚠️ Typo in the filename (caps) - could cause issues with git and/or MacOS.

pierre commented on catalog/src/main/resources/org/killbill/billing/catalog/ddl.sql:
```diff
@@ -41,3 +41,99 @@ CREATE TABLE catalog_override_plan_phase (
+create table catalog_override_tier_block
+(
+block_number smallint(5) unsigned,
+record_id serial unique,
```
`record_id` should be the first column.

sruthipendyala replied:
Yes, Will fix this!

sbrossie commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/json/PhasePriceOverrideJson.java:
```diff
@@ -123,14 +177,36 @@ public int hashCode() {
             public PlanPhasePriceOverride apply(@Nullable final PhasePriceOverrideJson input) {
+
+                List <UsagePriceOverride> usagePriceOverrides = new LinkedList<UsagePriceOverride>();
+                assert input != null;
```
We usually use `Preconditions.checkNotNull` instead of `assert`.

sruthipendyala commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultPlanPhase.java:
```diff
@@ -75,7 +77,19 @@ public DefaultPlanPhase(final DefaultPlan parentPlan, final DefaultPlanPhase in,
+                     return input !=null && input.getName().equals(curUsage.getName());
+                 }
+                 }).orNull();
+                usages[i] = (usagePriceOverride !=null) ? new DefaultUsage(in.getUsages()[i], usagePriceOverride, override.getCurrency()) : (DefaultUsage)curUsage;
```
@sbrossie : I am validating here if any override exists for the usage, and if it does I am replacing it with the override. This would work right?

sruthipendyala commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultPlanPhase.java:
```diff
@@ -75,7 +77,19 @@ public DefaultPlanPhase(final DefaultPlan parentPlan, final DefaultPlanPhase in,
+            if(override != null && override.getUsagePriceOverrides()!= null) {
+                final UsagePriceOverride usagePriceOverride = Iterables.tryFind(override.getUsagePriceOverrides(), new Predicate<UsagePriceOverride>() {
+                 @Override
+                 public boolean apply(final UsagePriceOverride input) {
```
@sbrossie : I am validating here to see if any override exists for the current usage based on the usage name and if it does I am replacing it with the override. This would work and take care of the ordering right?

sruthipendyala commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultTier.java:
```diff
@@ -62,6 +67,31 @@ public DefaultTier() {
+                    }
+
+                }).orNull();
+                blocks[i] = (overriddenTierBlock != null) ? new DefaultTieredBlock(in.getTieredBlocks()[i], overriddenTierBlock, currency) :
```
Validating using tiered block's name size and max values to see if any override exists for the current tier block. Same thought as the usage override. Will this not work?

sbrossie replied:
Yep, i think you are correct. I misread the code. Please ignore the remark

sruthipendyala commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultBlock.java:
```diff
@@ -17,23 +17,25 @@
 
 package org.killbill.billing.catalog;
 
-import javax.xml.bind.annotation.XmlAccessType;
```
Did not know about this. Yes will definitely follow the style from now on!!

sbrossie commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultInternationalPrice.java:
```diff
@@ -65,6 +65,18 @@ public DefaultInternationalPrice(final DefaultInternationalPrice in, final PlanP
+        for (int i = 0; i < in.getPrices().length; i++) {
+            final DefaultPrice curPrice = (DefaultPrice)  in.getPrices()[i];
+            if (curPrice.getCurrency().equals(currency)){
+                prices[i] = new DefaultPrice(overriddenPrice, currency);
```
Ignore remark, looks fine.

sruthipendyala commented:
@sbrossie : I worked on fixing the failing tests and also making some code changes based on the review. Added copyright information and changed the way in which the classes are imported.

I have started writing DAO level tests and will submit a pull request soon.

pierre commented:
@sruthipendyala I've merged your work in the work-for-release-0.19.x branch (see https://github.com/killbill/killbill/pull/665), so it's easier to review moving forward (smaller PRs).

Thanks!
